src/analysis/inner_product_space/of_norm.lean

```diff
@@ -0,0 +1,346 @@
+/-
+Copyright (c) 2020 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+
+import topology.algebra.algebra
+import analysis.inner_product_space.basic
+
+/-!
+# Inner product space derived from a norm
+
+This file defines an `inner_product_space` instance from a norm that respects the
+parallellogram identity. The parallelogram identity is a way to express the inner product of `x` and
+`y` in terms of the norms of `x`, `y`, `x + y`, `x - y`.
+
+## Main results
+
+- `inner_product_space.of_norm`: a normed space whose norm respects the parallellogram identity,
+  can be seen as an inner product space.
+
+## Tags
+
+inner product space, Hilbert space, norm
+
+## References
+
+- http://www.mathematik.uni-muenchen.de/~michel/jordan-von_neumann_-_parallelogram_identity.pdf
+- https://math.stackexchange.com/questions/21792/norms-induced-by-inner-products-and-the-parallelogram-law
+- https://math.dartmouth.edu/archive/m113w10/public_html/jordan-vneumann-thm.pdf
+-/
+
+variables {𝕜 E α : Type*} [is_R_or_C 𝕜] [normed_add_comm_group E] [normed_space 𝕜 E]
+  [topological_space α]
+
+local notation `𝓚` := algebra_map ℝ 𝕜
+open is_R_or_C
+open_locale complex_conjugate
+
+variables (𝕜)
+
+/-- Auxiliary definition of the inner product derived from the norm. -/
+private noncomputable def inner_ (x y : E) : 𝕜 :=
+4⁻¹ * ((𝓚 ‖x + y‖) * (𝓚 ‖x + y‖) - (𝓚 ‖x - y‖) * (𝓚 ‖x - y‖)
+          + (I:𝕜) * (𝓚 ‖(I:𝕜) • x + y‖) * (𝓚 ‖(I:𝕜) • x + y‖)
+          - (I:𝕜) * (𝓚 ‖(I:𝕜) • x - y‖) * (𝓚 ‖(I:𝕜) • x - y‖))
+
+variables (E)
+
+/-- Predicate for the parallelogram identity to hold. -/
+def parallelogram_identity : Prop :=
+∀ x y : E, ‖x + y‖ * ‖x + y‖ + ‖x - y‖ * ‖x - y‖ = 2 * (‖x‖ * ‖x‖ + ‖y‖ * ‖y‖)
```

Comment:
Should this be a typeclass?

Reply:
Like `inner_product_spaceable`, aka "`inner_product_space`, but Prop"?

Reply:
I'd wait for a use case.

Reply:
`quaternion` is a use-case

Reply:
One compelling use case is that it prevents lemmas having to specify the base field over which they are an inner-product space, which makes them less annoying to apply.

Reply:
Which current lemmas can you think of that require `inner_product_space` but don't mention `inner` in their type? The only ones I can find at the moment are
* the different versions of the parallelogram identity
* [`inner_product_space.to_uniform_convex_space`](https://leanprover-community.github.io/mathlib_docs/analysis/inner_product_space/basic.html#inner_product_space.to_uniform_convex_space)

Reply:
There are a few lemmas about continuity/differentiability of norms that I believe only need the inner product in the proof